benchmark/benchmark.py

```diff
@@ -0,0 +1,659 @@
+        else:
+            args.big_model = False
+        args.batch_size = batch_size
+        args.n_epochs = epochs
+        args.parallel = parallel.to_parallelism_type()
+        args.sharding = parallel.to_sharding_strategy()
+        fn = parallel_orbit_5k.main
+    elif model in (Models.bert, Models.bertbig):
+        if model is Models.bertbig:
+            args.big_model = True
+        else:
+            args.big_model = False
+        args.batch_size = batch_size
+        args.n_epochs = epochs
+        args.parallel = parallel.to_parallelism_type()
+        args.sharding = parallel.to_sharding_strategy()
+        args.download = False
+        parallel_bert.download_dataset()
+        fn = parallel_bert.main
+
+    sys.stdout.write(
+        "++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n"
+    )
+    sys.stdout.write(
+        f"BENCHMARK RUNNING ON {device_name}... parallelism {parallel} with batch size {batch_size}...\n"
+    )
+    sys.stdout.flush()
+
+    # Spawn a new python interpreter to ensure a nice release of resources
+    mp = pmp.get_context("spawn")
+    rq = mp.SimpleQueue()
+    process = mp.Process(target=wrap, args=(fn, args, rq), daemon=False)
+    process.start()
+    process.join()
+    if process.exitcode != 0:
+        raise Exception(f"Train process exited with exitcode {process.exitcode}")
+    r = rq.get()
+
+    # if model is not Models.none and parallel is Parallelism.pipeline:
+    #    torch.distributed.rpc.shutdown()
+
+    return RunData(
+        r.start_time,
+        r.end_time,
+        model,
+        parallel,
+        epochs,
+        batch_size,
+        r.loss,
+        r.accuracy,
+        device_count,
+        device_model,
+    )
+
+
+def uniq(data: typing.List[RunData]):
+    """Keep the most recent elements of each class.
+
+    Some elements of the list may be of the same class but of different generation
+    time, e.g. some benchmark runs that were restarted.
+    """
+    data2 = []
+    idx = 0
+    # parse every element in the list (unless those removed during the process)
+    while idx < len(data):
+        jdx = idx + 1
+        keep = data[idx]  # set current data as kept
+        # parse every further element in the list (unless those removed during the process)
+        while jdx < len(data):
+            # if the currently kept element and the current element are of the same "class" ...
+            if data[jdx].same(keep):
+                # ... compare if the current element is greater than the kept one ...
+                if data[jdx].gt(keep):
+                    # ... and keep and remove the current element if it is greater
+                    keep = data.pop(jdx)
+                else:
+                    # ... or only remove the current element if it is not greater
+                    del data[jdx]
+            else:
+                jdx += 1
+        data2.append(keep)
+        idx += 1
+    return data2
```

Comment:
The naming is terrible. Why is `uniq` not `unique` or, better, a more descriptive name? Also, `data2` doesn't say anything. The logic is super complicated: two nested while loops with nested if-else statements, and the comments are just repeating what the code already expresses.

This can be written more concisely as:
```
def get_unique_latest_runs(data: typing.List[RunData]) -> typing.List[RunData]:
    unique_config_latest_run = {}
   
    for run in data:
        configuration_key = (run.model, run.parallel, run.batch_size, run.gpu_count, run.gpu_model)
        if configuration_key not in unique_config_latest_run or run.end_time > unique_config_latest_run[configuration_key].end_time:
            unique_config_latest_run[configuration_key] = run
   
    return list(unique_config_latest_run.values())
```